pkg/middleware/http/ratelimiter.go

```diff
@@ -0,0 +1,199 @@
+package http
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	slimiter "github.com/ulule/limiter/v3"
+	smemory "github.com/ulule/limiter/v3/drivers/store/memory"
+
+	"github.com/bnb-chain/greenfield-storage-provider/pkg/log"
+)
+
+type RateLimiterCell struct {
+	Key        string
+	RateLimit  int
+	RatePeriod string
+}
+
+type HTTPLimitConfig struct {
+	On         bool
+	RateLimit  int
+	RatePeriod string
+}
+
+type RateLimiterConfig struct {
+	HTTPLimitCfg HTTPLimitConfig
+	PathPattern  []RateLimiterCell
+	HostPattern  []RateLimiterCell
+	APILimits    []RateLimiterCell
+}
+
+type MemoryLimiterConfig struct {
+	RateLimit  int    // rate
+	RatePeriod string // per period
+}
+
+type APILimiterConfig struct {
```

Comment:
APILimiterConfig does not only include APILimits. Configuration semantics are not particularly clear.